{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41c8b971-d2e6-4852-9be9-9497cbbef359",
   "metadata": {},
   "outputs": [],
   "source": [
    "import micom\n",
    "import micom.measures\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import scipy.stats\n",
    "from matplotlib.patches import Rectangle\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from plotnine import *\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bb69feb-309b-4972-acc7-fbb10104a306",
   "metadata": {},
   "source": [
    "## Load Taxonomic Data\n",
    "Arivale metagenomes mapped to taxonomy with Kraken2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4fe07762-e914-4728-8545-1b835de18a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy = pd.read_csv('../data/arivale-metagenomes.csv')\n",
    "taxonomy['sample_id'] = taxonomy['sample_id'].astype('str')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb3d8c50-f72a-4dbd-8e8e-83d49b884053",
   "metadata": {},
   "source": [
    "## Add Simulated Probiotics\n",
    "Add five-species probiotic cocktail to each Arivale sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "38684de0-5c7e-439e-a73a-11f6350683a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_all_probiotic(taxonomy):\n",
    "    \"\"\"Adds probiotic cocktail of AMUC, CBEI, CBUT, BINF, and EHAL to existing taxonomy table at 1% RA each\"\"\"\n",
    "    taxonomy_probiotic_total = pd.DataFrame()\n",
    "    for sample in taxonomy['sample_id'].unique():\n",
    "        taxonomy_reduced = taxonomy[taxonomy['sample_id'] == sample].copy()\n",
    "        taxonomy_reduced['abundance'] = taxonomy_reduced['abundance']*0.95\n",
    "        taxonomy_probiotic = pd.concat([taxonomy_reduced,pd.DataFrame({\n",
    "                  'sample_id':[sample, sample, sample, sample, sample], \n",
    "                  'species': ['Akkermansia muciniphila','Clostridium beijerinckii',\n",
    "                              'Clostridium butyricum','Bifidobacterium longum',\n",
    "                              'Anaerobutyricum hallii'],\n",
    "                  'abundance':[0.01, 0.01, 0.01, 0.01, 0.01], \n",
    "                  'id':['Akkermansia_muciniphila','Clostridium_beijerinckii',\n",
    "                              'Clostridium_butyricum','Bifidobacterium_longum',\n",
    "                              'Anaerobutyricum_hallii']})])\n",
    "        taxonomy_probiotic_total = pd.concat([taxonomy_probiotic_total, taxonomy_probiotic])\n",
    "\n",
    "    return taxonomy_probiotic_total\n",
    "# Add probiotics\n",
    "taxonomy_probiotic = add_all_probiotic(taxonomy)\n",
    "\n",
    "# Combine duplicate taxa\n",
    "taxonomy_probiotic = taxonomy_probiotic.groupby(['sample_id','id','species']).sum().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97e5588b-b011-4753-a53c-88d8c37d0dfc",
   "metadata": {},
   "source": [
    "## Build Models\n",
    "Build models for all samples in the Arivale cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "596cf715-6d5b-47c8-8fe8-aefa5c21677e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "da7b58bfcfd642a5babe311ed6f59d64",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2;36m[06/17/25 03:08:42]\u001b[0m\u001b[2;36m \u001b[0m\u001b[33mWARNING \u001b[0m Less than \u001b[1;36m50\u001b[0m% of the abundance     \u001b]8;id=248784;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py\u001b\\\u001b[2mcommunity.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=389187;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py#229\u001b\\\u001b[2m229\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m                    \u001b[0m         could be matched to the model      \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         database. Model `\u001b[1;36m22001612560204\u001b[0m`   \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         may not be representative of the   \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         sample                             \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m[06/17/25 03:08:42]\u001b[0m\u001b[2;36m \u001b[0m\u001b[33mWARNING \u001b[0m Less than \u001b[1;36m50\u001b[0m% of the abundance     \u001b]8;id=340494;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py\u001b\\\u001b[2mcommunity.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=792257;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py#229\u001b\\\u001b[2m229\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m                    \u001b[0m         could be matched to the model      \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         database. Model `\u001b[1;36m22001612560096\u001b[0m`   \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         may not be representative of the   \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         sample                             \u001b[2m                \u001b[0m\n"
     ]
    }
   ],
   "source": [
    "agora = '../agora_models/data'\n",
    "manifest_treatment = micom.workflows.build(taxonomy_probiotic,\n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../Arivale_treatment_models',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n",
    "manifest_baseline = micom.workflows.build(taxonomy,\n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../Arivale_baseline_models',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61da7f8b-48d6-4aaf-a426-4bee40036827",
   "metadata": {},
   "source": [
    "## Prepare media\n",
    "Load the standard European medium, and augment with addition fiber prebiotics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "be012a69-dcc7-4625-9a85-e73853ceaa77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load European medium\n",
    "eu_medium = micom.qiime_formats.load_qiime_medium('../european_medium.qza')\n",
    "\n",
    "# Make copies for adding prebiotics\n",
    "eu_medium_inulin = eu_medium.copy()\n",
    "eu_medium_pectin = eu_medium.copy()\n",
    "eu_medium_malto = eu_medium.copy()\n",
    "eu_medium_starch = eu_medium.copy()\n",
    "eu_medium_hemp = eu_medium.copy()\n",
    "eu_medium_psyllium= eu_medium.copy()\n",
    "\n",
    "# Add prebiotics\n",
    "eu_medium_inulin.loc['EX_inulin_m', 'flux'] = 6.14\n",
    "eu_medium_pectin.loc['EX_pect_m', 'flux'] = 0.4\n",
    "eu_medium_starch.loc['EX_strch1_m', 'flux'] = 16.65\n",
    "eu_medium_malto.loc['EX_dextrin_m', 'flux'] = 30.28\n",
    "eu_medium_hemp.loc['EX_cellul_m', 'flux'] = 0.37\n",
    "eu_medium_psyllium.loc['EX_arabinoxyl_m', 'flux'] = 4.09\n",
    "\n",
    "# Load high fiber medium\n",
    "hf_medium = micom.qiime_formats.load_qiime_medium('../high_fiber_medium.qza')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7276b652-6ba8-41db-89a2-2f4b06e1bea4",
   "metadata": {},
   "source": [
    "## Grow samples\n",
    "Grow samples for both probiotic treated and untreated samples using each medium. This will take a while!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3803796c-90c0-4a3c-aad5-89f583f9f175",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../Arivale_treament_models/manifest.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m manifest_treatment \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m../Arivale_treament_models/manifest.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m treatment_noPrebiotic \u001b[38;5;241m=\u001b[39m micom\u001b[38;5;241m.\u001b[39mworkflows\u001b[38;5;241m.\u001b[39mgrow(manifest_treatment, \n\u001b[1;32m      3\u001b[0m                                   model_folder \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../Arivale_treament_models/\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      4\u001b[0m                                   medium \u001b[38;5;241m=\u001b[39m eu_medium,\n\u001b[1;32m      5\u001b[0m                                   tradeoff \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.99\u001b[39m, \n\u001b[1;32m      6\u001b[0m                                   strategy \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m,                                    \n\u001b[1;32m      7\u001b[0m                                   threads \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m20\u001b[39m)\n\u001b[1;32m      8\u001b[0m treatment_inulin \u001b[38;5;241m=\u001b[39m micom\u001b[38;5;241m.\u001b[39mworkflows\u001b[38;5;241m.\u001b[39mgrow(manifest_treatment, \n\u001b[1;32m      9\u001b[0m                                   model_folder \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../Arivale_treament_models/\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m     10\u001b[0m                                   medium \u001b[38;5;241m=\u001b[39m eu_medium_inulin,\n\u001b[1;32m     11\u001b[0m                                   tradeoff \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.99\u001b[39m, \n\u001b[1;32m     12\u001b[0m                                   strategy \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m,                                    \n\u001b[1;32m     13\u001b[0m                                   threads \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m20\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/micom-host/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/micom-host/lib/python3.9/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/envs/micom-host/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/micom-host/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1880\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1879\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1880\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1883\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1884\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1885\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1886\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1887\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1888\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1889\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1890\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1891\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m~/miniconda3/envs/micom-host/lib/python3.9/site-packages/pandas/io/common.py:873\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    868\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    869\u001b[0m     \u001b[38;5;66;03m# Check whether the filename is to be opened in binary mode.\u001b[39;00m\n\u001b[1;32m    870\u001b[0m     \u001b[38;5;66;03m# Binary mode does not support 'encoding' and 'newline'.\u001b[39;00m\n\u001b[1;32m    871\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mencoding \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m ioargs\u001b[38;5;241m.\u001b[39mmode:\n\u001b[1;32m    872\u001b[0m         \u001b[38;5;66;03m# Encoding\u001b[39;00m\n\u001b[0;32m--> 873\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\n\u001b[1;32m    874\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    875\u001b[0m \u001b[43m            \u001b[49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    876\u001b[0m \u001b[43m            \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    877\u001b[0m \u001b[43m            \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    878\u001b[0m \u001b[43m            \u001b[49m\u001b[43mnewline\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    879\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    880\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    881\u001b[0m         \u001b[38;5;66;03m# Binary mode\u001b[39;00m\n\u001b[1;32m    882\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mopen\u001b[39m(handle, ioargs\u001b[38;5;241m.\u001b[39mmode)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../Arivale_treament_models/manifest.csv'"
     ]
    }
   ],
   "source": [
    "manifest_treatment = pd.read_csv('../Arivale_treament_models/manifest.csv')\n",
    "treatment_noPrebiotic = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = eu_medium,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "treatment_inulin = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = eu_medium_inulin,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "treatment_pect = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = eu_medium_pectin,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "treatment_starch = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = eu_medium_starch,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "treatment_malto = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = eu_medium_malto,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "treatment_hemp = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = eu_medium_hemp,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "treatment_psyllium = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = eu_medium_psyllium,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "treatment_hf = micom.workflows.grow(manifest_treatment, \n",
    "                                  model_folder = '../Arivale_treament_models/',\n",
    "                                  medium = hf_medium,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "\n",
    "\n",
    "manifest_baseline = pd.read_csv('../Arivale_baseline_models/manifest.csv')\n",
    "baseline_noPrebiotic = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = eu_medium,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "baseline_inulin = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = eu_medium_inulin,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "baseline_pect = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = eu_medium_pectin,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "baseline_starch = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = eu_medium_starch,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "baseline_malto = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = eu_medium_malto,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "baseline_hemp = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = eu_medium_hemp,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "baseline_psyllium = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = eu_medium_psyllium,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)\n",
    "baseline_hf = micom.workflows.grow(manifest_baseline, \n",
    "                                  model_folder = '../Arivale_baseline_models/',\n",
    "                                  medium = hf_medium,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',                                    \n",
    "                                  threads = 20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ab65063-6c35-46d4-b7d1-3716c0f3d452",
   "metadata": {},
   "source": [
    "## Collect Growth Rates\n",
    "Return the MCMM-predicted growth rates for each prebiotic/probiotic context"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b727de4e-091e-4b09-9978-fcd6c134d36d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather growth rates from treatment samples\n",
    "rates_noPrebiotic = treatment_noPrebiotic.growth_rates\n",
    "rates_noPrebiotic['treatment'] = 'No Prebiotic'\n",
    "rates_noPrebiotic['probiotic'] = '+'\n",
    "rates_inulin = treatment_inulin.growth_rates\n",
    "rates_inulin['treatment'] = 'Inulin'\n",
    "rates_inulin['probiotic'] = '+'\n",
    "rates_pect = treatment_pect.growth_rates\n",
    "rates_pect['treatment'] = 'Pectin'\n",
    "rates_pect['probiotic'] = '+'\n",
    "rates_starch = treatment_starch.growth_rates\n",
    "rates_starch['treatment'] = 'Starch'\n",
    "rates_starch['probiotic'] = '+'\n",
    "rates_malto = treatment_malto.growth_rates\n",
    "rates_malto['treatment'] = 'Maltodextrin'\n",
    "rates_malto['probiotic'] = '+'\n",
    "rates_hemp = treatment_hemp.growth_rates\n",
    "rates_hemp['treatment'] = 'Hemp Seed'\n",
    "rates_hemp['probiotic'] = '+'\n",
    "rates_psyllium = treatment_psyllium.growth_rates\n",
    "rates_psyllium['treatment'] = 'Psyllium Husk'\n",
    "rates_psyllium['probiotic'] = '+'\n",
    "rates_hf = treatment_hf.growth_rates\n",
    "rates_hf['treatment'] = 'High Fiber'\n",
    "rates_hf['probiotic'] = '+'\n",
    "\n",
    "# Baseline samples\n",
    "baseline_rates_noPrebiotic = baseline_noPrebiotic.growth_rates\n",
    "baseline_rates_noPrebiotic['treatment'] = 'No Prebiotic'\n",
    "baseline_rates_noPrebiotic['probiotic'] = '-'\n",
    "baseline_rates_inulin = baseline_inulin.growth_rates\n",
    "baseline_rates_inulin['treatment'] = 'Inulin'\n",
    "baseline_rates_inulin['probiotic'] = '-'\n",
    "baseline_rates_pect = baseline_pect.growth_rates\n",
    "baseline_rates_pect['treatment'] = 'Pectin'\n",
    "baseline_rates_pect['probiotic'] = '-'\n",
    "baseline_rates_starch = baseline_starch.growth_rates\n",
    "baseline_rates_starch['treatment'] = 'Starch'\n",
    "baseline_rates_starch['probiotic'] = '-'\n",
    "baseline_rates_malto = baseline_malto.growth_rates\n",
    "baseline_rates_malto['treatment'] = 'Maltodextrin'\n",
    "baseline_rates_malto['probiotic'] = '-'\n",
    "baseline_rates_hemp = baseline_hemp.growth_rates\n",
    "baseline_rates_hemp['treatment'] = 'Hemp Seed'\n",
    "baseline_rates_hemp['probiotic'] = '-'\n",
    "baseline_rates_psyllium = baseline_pysllium.growth_rates\n",
    "baseline_rates_psyllium['treatment'] = 'Psyllium Husk'\n",
    "baseline_rates_psyllium['probiotic'] = '-'\n",
    "baseline_rates_hf = baseline_hf.growth_rates\n",
    "baseline_rates_hf['treatment'] = 'High Fiber'\n",
    "baseline_rates_hf['probiotic'] = '-'\n",
    "\n",
    "\n",
    "# Concatenate growth rates into a single DF\n",
    "rates_probiotic = pd.concat([rates_noPrebiotic, rates_inulin, rates_pect, rates_starch, rates_malto, rates_hemp, rates_psyllium, rates_hf,\n",
    "                            baseline_rates_noPrebiotic, baseline_rates_inulin, baseline_rates_pect, baseline_rates_starch, baseline_rates_malto,baseline_rates_hemp,baseline_rates_psyllium,\n",
    "                             baseline_rates_hf])\n",
    "\n",
    "\n",
    "\n",
    "# Isolate Probiotics\n",
    "rates_probiotic = rates_probiotic[rates_probiotic.index.isin(['Akkermansia_muciniphila',\n",
    "                                                              'Clostridium_beijerinckii',\n",
    "                                                              'Clostridium_butyricum',\n",
    "                                                              'Bifidobacterium_longum',\n",
    "                                                              'Anaerobutyricum_hallii',\n",
    "                                                              ])].reset_index()\n",
    "\n",
    "\n",
    "# Abbreviate probiotics for clarity\n",
    "rates_probiotic.compartments = rates_probiotic.compartments.str.replace('Akkermansia_muciniphila','AMUC').str.replace(\n",
    "                                                          'Clostridium_beijerinckii','CBEI').str.replace(\n",
    "                                                          'Clostridium_butyricum','CBUT').str.replace(\n",
    "                                                          'Bifidobacterium_longum','BINF').str.replace(\n",
    "                                                          'Anaerobutyricum_hallii', 'AHAL')\n",
    "\n",
    "# Create unique sample ID for each treatment combination\n",
    "rates_probiotic['sample_id'] = rates_probiotic['sample_id'].astype('str')+'_'+rates_probiotic['treatment']+'_'+rates_probiotic['probiotic']\n",
    "\n",
    "# Log scale growth rates\n",
    "rates_probiotic['growth_rate_scaled'] = np.log10(rates_probiotic[['growth_rate']]+0.00000001)\n",
    "\n",
    "# Threshold log-scaled growth rates\n",
    "rates_probiotic.loc[rates_probiotic['growth_rate_scaled']>=-3, 'MCMM_engraftment'] = 1\n",
    "rates_probiotic['MCMM_engraftment'].fillna(0.0, inplace = True)\n",
    "\n",
    "# Pivot data so rows are samples and columns and probiotic species\n",
    "rates_probiotic_pvt = pd.pivot_table(rates_probiotic, \n",
    "                    columns='compartments',\n",
    "                    index='sample_id',\n",
    "                    values='MCMM_engraftment').fillna(0.0)\n",
    "\n",
    "# Format for plotting\n",
    "rates_probiotic_pvt['treatment'] = rates_probiotic_pvt.index.str.split('_').str[1]\n",
    "rates_probiotic_pvt['probiotic'] = rates_probiotic_pvt.index.str.split('_').str[2]\n",
    "rates_probiotic_pvt = rates_probiotic_pvt.sort_values(by = 'treatment')\n",
    "\n",
    "# Order and set row colors\n",
    "custom_order = ['No Prebiotic','Inulin','Pectin','Starch','Maltodextrin','Hemp Seed','Psyllium Husk','High Fiber']\n",
    "rates_probiotic_pvt['treatment'] = pd.Categorical(rates_probiotic_pvt['treatment'], categories=custom_order, ordered=True)\n",
    "rates_probiotic_pvt = rates_probiotic_pvt.sort_values('treatment')\n",
    "\n",
    "palette = sns.color_palette(\"Set2\", 8)\n",
    "sns.set(font_scale=3)\n",
    "lut = dict(zip(rates_probiotic_pvt['treatment'].unique(),palette))\n",
    "row_colors = rates_probiotic_pvt['treatment'].map(lut)\n",
    "\n",
    "# Save output\n",
    "rates_probiotic.to_csv('../arivale_probiotic_rates.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2bfef2-0ca9-400d-b64a-d4dd5e61398b",
   "metadata": {},
   "source": [
    "## Plot Engraftment of Probiotics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e6b4f7-c681-4290-9aa2-cbf213137609",
   "metadata": {},
   "outputs": [],
   "source": [
    "x = sns.clustermap(rates_probiotic_pvt[rates_probiotic_pvt['probiotic'] == '+'].iloc[:,:-2].T, yticklabels=True, \n",
    "                   col_colors = row_colors, col_cluster = False, xticklabels=False,  cmap='gray_r',cbar_pos=None,   # <- This removes the color bar\n",
    "                   figsize=(30, 8), dendrogram_ratio=0.05)\n",
    "\n",
    "for label in rates_probiotic_pvt['treatment'].unique():\n",
    "    x.ax_col_dendrogram.bar(0, 0, color=lut[label],\n",
    "                            label=label, linewidth=0)\n",
    "x.ax_col_dendrogram.legend(loc='upper center', bbox_to_anchor=(0.5, 6), ncol=4)\n",
    "x.savefig('../figures/arivale_heatmap.svg', dpi=200)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63c1ddea-8b05-4c67-80c0-5ba883b00156",
   "metadata": {},
   "source": [
    "## Calculate Production Rates\n",
    "Return the MCMM-predicted metabolic production for each prebiotic/probiotic combination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e24352a-e96f-466e-b288-b7ef9437e98e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather production rates from treatment samples\n",
    "production_noPrebiotic = micom.measures.production_rates(treatment_noPrebiotic)\n",
    "production_noPrebiotic['treatment'] = 'No Prebiotic'\n",
    "production_noPrebiotic['probiotic'] = '+'\n",
    "\n",
    "production_inulin = micom.measures.production_rates(treatment_inulin)\n",
    "production_inulin['treatment'] = 'Inulin'\n",
    "production_inulin['probiotic'] = '+'\n",
    "\n",
    "production_pect = micom.measures.production_rates(treatment_pect)\n",
    "production_pect['treatment'] = 'Pectin'\n",
    "production_pect['probiotic'] = '+'\n",
    "\n",
    "production_starch = micom.measures.production_rates(treatment_starch)\n",
    "production_starch['treatment'] = 'Starch'\n",
    "production_starch['probiotic'] = '+'\n",
    "\n",
    "production_malto = micom.measures.production_rates(treatment_malto)\n",
    "production_malto['treatment'] = 'Maltodextrin'\n",
    "production_malto['probiotic'] = '+'\n",
    "\n",
    "production_hemp = micom.measures.production_rates(treatment_hemp)\n",
    "production_hemp['treatment'] = 'Hemp Seed'\n",
    "production_hemp['probiotic'] = '+'\n",
    "\n",
    "production_psyllium = micom.measures.production_rates(treatment_psyllium)\n",
    "production_psyllium['treatment'] = 'Psyllium Husk'\n",
    "production_psyllium['probiotic'] = '+'\n",
    "\n",
    "production_hf = micom.measures.production_rates(treatment_hf)\n",
    "production_hf['treatment'] = 'High Fiber'\n",
    "production_hf['probiotic'] = '+'\n",
    "\n",
    "# Baseline samples\n",
    "baseline_production_noPrebiotic = micom.measures.production_rates(baseline_noPrebiotic)\n",
    "baseline_production_noPrebiotic['treatment'] = 'No Prebiotic'\n",
    "baseline_production_noPrebiotic'probiotic'] = '-'\n",
    "\n",
    "baseline_production_inulin = micom.measures.production_rates(baseline_inulin)\n",
    "baseline_production_inulin['treatment'] = 'Inulin'\n",
    "baseline_production_inulin['probiotic'] = '-'\n",
    "\n",
    "baseline_production_pect = micom.measures.production_rates(baseline_pect)\n",
    "baseline_production_pect['treatment'] = 'Pectin'\n",
    "baseline_production_pect['probiotic'] = '-'\n",
    "\n",
    "baseline_production_starch = micom.measures.production_rates(baseline_starch)\n",
    "baseline_production_starch['treatment'] = 'Starch'\n",
    "baseline_production_starch['probiotic'] = '-'\n",
    "\n",
    "baseline_production_malto = micom.measures.production_rates(baseline_malto)\n",
    "baseline_production_malto['treatment'] = 'Maltodextrin'\n",
    "baseline_production_malto['probiotic'] = '-'\n",
    "\n",
    "baseline_production_hemp = micom.measures.production_rates(baseline_hemp)\n",
    "baseline_production_hemp['treatment'] = 'Hemp Seed'\n",
    "baseline_production_hemp['probiotic'] = '-'\n",
    "\n",
    "baseline_production_psyllium = micom.measures.production_rates(baseline_psyllium)\n",
    "baseline_production_psyllium['treatment'] = 'Psyllium Husk'\n",
    "baseline_production_psyllium['probiotic'] = '-'\n",
    "\n",
    "\n",
    "baseline_production_hf = micom.measures.production_rates(baseline_hf)\n",
    "baseline_production_hf['treatment'] = 'High Fiber'\n",
    "baseline_production_hf['probiotic'] = '-'\n",
    "\n",
    "# Concatenate results\n",
    "production = pd.concat([production_baseline, production_inulin, production_pect, production_starch, production_malto, production_hemp, production_psyllium, production_hf,\n",
    "                       baseline_production_baseline, baseline_production_inulin, baseline_production_pect, baseline_production_starch, baseline_production_malto,\n",
    "                        baseline_production_hemp, baseline_production_psyllium, baseline_production_hf])\n",
    "\n",
    "# Isolate butyrate and propionate\n",
    "production_but = production[production['metabolite'] == 'but[e]']\n",
    "production_ppa = production[production['metabolite'] == 'ppa[e]']\n",
    "production_but\n",
    "\n",
    "production.to_csv('../arivale_metabolic_production.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "665697ff-7f68-426d-930e-b110313963b4",
   "metadata": {},
   "source": [
    "## Plot Butyrate Production\n",
    "Boxplots display butyrate production between prebiotic and probiotic contexts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "051e95f9-da88-401b-91dc-511f1c03c9bb",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'production_but' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m arivale_butyrate \u001b[38;5;241m=\u001b[39m (ggplot(\u001b[43mproduction_but\u001b[49m, aes(x \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtreatment\u001b[39m\u001b[38;5;124m'\u001b[39m, y \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mflux\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m      2\u001b[0m                     \u001b[38;5;241m+\u001b[39mgeom_boxplot(aes(fill \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprobiotic\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m      3\u001b[0m                     \u001b[38;5;241m+\u001b[39mscale_fill_manual(values \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#5B85AA\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      4\u001b[0m                                                  \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#F83860\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      5\u001b[0m                                                 })\n\u001b[1;32m      6\u001b[0m                     \u001b[38;5;241m+\u001b[39mscale_x_discrete(limits \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNo Prebiotic\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mInulin\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPectin\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mStarch\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMaltodextrin\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPsyllium Husk\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHemp Seed\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHigh Fiber\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m      7\u001b[0m                     \u001b[38;5;241m+\u001b[39mlabs(y \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPredicted Butyate \u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m mmol/gDW/h\u001b[39m\u001b[38;5;124m'\u001b[39m, x \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m,title \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mButyrate\u001b[39m\u001b[38;5;124m'\u001b[39m, fill \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mProbiotic Treatment\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      8\u001b[0m                     \u001b[38;5;241m+\u001b[39mtheme_minimal()    \n\u001b[1;32m      9\u001b[0m                     \u001b[38;5;241m+\u001b[39mylim(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m100\u001b[39m)\n\u001b[1;32m     10\u001b[0m                     \u001b[38;5;241m+\u001b[39mtheme(figure_size\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m8\u001b[39m, \u001b[38;5;241m8\u001b[39m),legend_position\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnone\u001b[39m\u001b[38;5;124m'\u001b[39m, text\u001b[38;5;241m=\u001b[39melement_text(size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m24\u001b[39m),axis_text_x\u001b[38;5;241m=\u001b[39melement_text(rotation\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m33\u001b[39m, hjust\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m))\n\u001b[1;32m     11\u001b[0m                    )\n\u001b[1;32m     14\u001b[0m ggsave(arivale_butyrate, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../figures/arivale_butyrate.svg\u001b[39m\u001b[38;5;124m'\u001b[39m, dpi \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m300\u001b[39m)\n\u001b[1;32m     16\u001b[0m arivale_butyrate\n",
      "\u001b[0;31mNameError\u001b[0m: name 'production_but' is not defined"
     ]
    }
   ],
   "source": [
    "arivale_butyrate = (ggplot(production_but, aes(x = 'treatment', y = 'flux'))\n",
    "                    +geom_boxplot(aes(fill = 'probiotic'))\n",
    "                    +scale_fill_manual(values = {'+': '#5B85AA',\n",
    "                                                 '-': '#F83860'\n",
    "                                                })\n",
    "                    +scale_x_discrete(limits = ['No Prebiotic', 'Inulin','Pectin','Starch','Maltodextrin','Psyllium Husk','Hemp Seed','High Fiber'])\n",
    "                    +labs(y = 'Predicted Butyate \\n mmol/gDW/h', x = '',title = 'Butyrate', fill = 'Probiotic Treatment')\n",
    "                    +theme_minimal()    \n",
    "                    +ylim(0, 100)\n",
    "                    +theme(figure_size=(8, 8),legend_position='none', text=element_text(size=24),axis_text_x=element_text(rotation=33, hjust=1))\n",
    "                   )\n",
    "\n",
    "\n",
    "ggsave(arivale_butyrate, '../figures/arivale_butyrate.svg', dpi = 300)\n",
    "\n",
    "arivale_butyrate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35a433fb-d1d0-4f2d-8da9-5e7401f4e49f",
   "metadata": {},
   "source": [
    "## Test Significance\n",
    "Use non-parametric Mann-Whitney to determine signficance in difference of propionate production between probiotic and baseline conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "44c13f73-01c3-4ff9-894e-5cd859253936",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'production_but' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m scipy\u001b[38;5;241m.\u001b[39mstats\u001b[38;5;241m.\u001b[39mmannwhitneyu(\u001b[43mproduction_but\u001b[49m[(production_but[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtreatment\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNo Prebiotic\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m&\u001b[39m(production_but[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprobiotic\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m'\u001b[39m)][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mflux\u001b[39m\u001b[38;5;124m'\u001b[39m],\n\u001b[1;32m      2\u001b[0m production_but[(production_but[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtreatment\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNo Prebiotic\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m&\u001b[39m(production_but[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprobiotic\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m)][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mflux\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "\u001b[0;31mNameError\u001b[0m: name 'production_but' is not defined"
     ]
    }
   ],
   "source": [
    "scipy.stats.mannwhitneyu(production_but[(production_but['treatment'] == 'No Prebiotic')&(production_but['probiotic']== '+')]['flux'],\n",
    "production_but[(production_but['treatment'] == 'No Prebiotic')&(production_but['probiotic']== '-')]['flux'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1d2cf82-76a4-4075-8ea8-1fa22eb2f5be",
   "metadata": {},
   "source": [
    "## Plot Propionate Production\n",
    "Boxplots display propionate production between prebiotic and probiotic contexts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "3a8f919c-b614-4fd1-b94c-a9d22344d78a",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'production_ppa' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m arivale_propionate \u001b[38;5;241m=\u001b[39m (ggplot(\u001b[43mproduction_ppa\u001b[49m, aes(x \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtreatment\u001b[39m\u001b[38;5;124m'\u001b[39m, y \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mflux\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m      2\u001b[0m                     \u001b[38;5;241m+\u001b[39mgeom_boxplot(aes(fill \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mprobiotic\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m      3\u001b[0m                     \u001b[38;5;241m+\u001b[39mtheme_seaborn()\n\u001b[1;32m      4\u001b[0m                     \u001b[38;5;241m+\u001b[39mscale_fill_manual(values \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#5B85AA\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      5\u001b[0m                                                  \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m#F83860\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      6\u001b[0m                                                 })\n\u001b[1;32m      7\u001b[0m                     \u001b[38;5;241m+\u001b[39mscale_x_discrete(limits \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNo Prebiotic\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mInulin\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPectin\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mStarch\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMaltodextrin\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPsyllium Husk\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHemp Seed\u001b[39m\u001b[38;5;124m'\u001b[39m,\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHigh Fiber\u001b[39m\u001b[38;5;124m'\u001b[39m])\n\u001b[1;32m      8\u001b[0m                     \u001b[38;5;241m+\u001b[39mlabs(y \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPredicted Propionate \u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m mmol/gDW/h\u001b[39m\u001b[38;5;124m'\u001b[39m, x \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m,title \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mPropionate\u001b[39m\u001b[38;5;124m'\u001b[39m, fill \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mProbiotic Treatment\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      9\u001b[0m                     \u001b[38;5;241m+\u001b[39mtheme_minimal()    \n\u001b[1;32m     10\u001b[0m                     \u001b[38;5;241m+\u001b[39mylim(\u001b[38;5;241m0\u001b[39m, \u001b[38;5;241m400\u001b[39m)\n\u001b[1;32m     11\u001b[0m                     \u001b[38;5;241m+\u001b[39mtheme(figure_size\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m12\u001b[39m, \u001b[38;5;241m8\u001b[39m), text\u001b[38;5;241m=\u001b[39melement_text(size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m24\u001b[39m),axis_text_x\u001b[38;5;241m=\u001b[39melement_text(rotation\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m33\u001b[39m, hjust\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m))\n\u001b[1;32m     12\u001b[0m                    )\n\u001b[1;32m     15\u001b[0m ggsave(arivale_propionate, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m../figures/arivale_propionate.svg\u001b[39m\u001b[38;5;124m'\u001b[39m, dpi \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m300\u001b[39m)\n\u001b[1;32m     17\u001b[0m arivale_propionate\n",
      "\u001b[0;31mNameError\u001b[0m: name 'production_ppa' is not defined"
     ]
    }
   ],
   "source": [
    "arivale_propionate = (ggplot(production_ppa, aes(x = 'treatment', y = 'flux'))\n",
    "                    +geom_boxplot(aes(fill = 'probiotic'))\n",
    "                    +theme_seaborn()\n",
    "                    +scale_fill_manual(values = {'+': '#5B85AA',\n",
    "                                                 '-': '#F83860'\n",
    "                                                })\n",
    "                    +scale_x_discrete(limits = ['No Prebiotic', 'Inulin','Pectin','Starch','Maltodextrin','Psyllium Husk','Hemp Seed','High Fiber'])\n",
    "                    +labs(y = 'Predicted Propionate \\n mmol/gDW/h', x = '',title = 'Propionate', fill = 'Probiotic Treatment')\n",
    "                    +theme_minimal()    \n",
    "                    +ylim(0, 400)\n",
    "                    +theme(figure_size=(12, 8), text=element_text(size=24),axis_text_x=element_text(rotation=33, hjust=1))\n",
    "                   )\n",
    "\n",
    "\n",
    "ggsave(arivale_propionate, '../figures/arivale_propionate.svg', dpi = 300)\n",
    "\n",
    "arivale_propionate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80349a0e-3af3-45cb-bfb7-ce5579442b43",
   "metadata": {},
   "source": [
    "## Test Significance\n",
    "Use non-parametric Mann-Whitney to determine signficance in difference of propionate production between conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b994bbc2-18fa-41bf-b368-9070164883c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.mannwhitneyu(production_ppa[(production_ppa['treatment'] == 'No Prebiotic')&(production_ppa['probiotic']== '+')]['flux'],\n",
    "production_ppa[(production_ppa['treatment'] == 'No Prebiotic')&(production_ppa['probiotic']== '-')]['flux'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6740deba-01c2-41e8-8a78-3da82f940c3e",
   "metadata": {},
   "source": [
    "## Format DF\n",
    "Reformat these data to allow us to find the most optimal intervention for each individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "381cbd77-0b61-4af8-87b3-d5a5a1bffa9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create unique ID for prebiotic/probiotic combos\n",
    "production_but['combined_treatment'] = production_but['treatment']+'_'+production_but['probiotic']\n",
    "production_ppa['combined_treatment'] = production_ppa['treatment']+'_'+production_ppa['probiotic']\n",
    "\n",
    "# Pivot so each row is a treatment, and each column is a sample\n",
    "production_but_pvt = pd.pivot_table(production_but, \n",
    "                                    columns = 'sample_id',\n",
    "                                    index = 'combined_treatment', \n",
    "                                    values = 'flux')\n",
    "production_ppa_pvt = pd.pivot_table(production_ppa, \n",
    "                                    columns = 'sample_id',\n",
    "                                    index = 'combined_treatment', \n",
    "                                    values = 'flux')\n",
    "\n",
    "# Remove high-fiber results\n",
    "production_but_pvt_noHF = production_but_pvt[~production_but_pvt.index.str.contains('High Fiber')]\n",
    "production_ppa_pvt_noHF = production_ppa_pvt[~production_ppa_pvt.index.str.contains('High Fiber')]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aea19a5-063f-4e57-8bac-769e8bea976f",
   "metadata": {},
   "source": [
    "# Find Optimal Intervention\n",
    "Isolate intervention that leads to maximal butyrate/propionate production in each individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5762bff-d06f-4171-9036-777aff4e4354",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create optimal propionate DF\n",
    "optimal_ppa = pd.DataFrame({'Optimal':production_ppa_pvt_noHF.max(),\n",
    "                            'Standard':production_ppa_pvt.loc['Inulin_+']}).reset_index()\n",
    "\n",
    "# Create optimal butyrate DF\n",
    "optimal_but = pd.DataFrame({'Optimal':production_but_pvt_noHF.max(),\n",
    "                            'Standard':production_but_pvt.loc['Inulin_+']}).reset_index()\n",
    "\n",
    "# Melt DFs \n",
    "optimal_ppa_melt = pd.melt(optimal_ppa, \n",
    "                id_vars = 'sample_id', \n",
    "                value_vars = ['Optimal','Standard'], \n",
    "                value_name = 'flux', \n",
    "                var_name = 'type')\n",
    "\n",
    "optimal_but_melt = pd.melt(optimal_but, \n",
    "                id_vars = 'sample_id', \n",
    "                value_vars = ['Optimal','Standard'], \n",
    "                value_name = 'flux', \n",
    "                var_name = 'type')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "082c9535-bb53-4b1e-9c81-2ef7b29b91fe",
   "metadata": {},
   "source": [
    "## Plot Optimal Propionate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc540d71-b783-450e-89b0-94e17a769e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "optimal_plot = (ggplot(optimal_ppa_melt, aes(x = 'type', y = 'flux'))\n",
    "                    +geom_boxplot(aes(fill = 'type'))\n",
    "                    +scale_fill_manual(values = {'Optimal': '#b2d0b1',\n",
    "                                                 'Standard': '#fee3b2'\n",
    "                                                })\n",
    "                    +labs(y = 'Predicted Propionate , mmol/gDW/h', x = '',title = 'Propionate', fill = 'Probiotic Treatment')\n",
    "                    +theme_minimal()    \n",
    "                    +theme(figure_size=(10, 5),\n",
    "                           legend_position='none',\n",
    "                           axis_title=element_text(size=18),\n",
    "                            axis_text=element_text(size=18),\n",
    "                            plot_title=element_text(size=20)))\n",
    "                \n",
    "               \n",
    "ggsave(optimal_plot, '../figures/optimal_propionate.svg',dpi =300)\n",
    "optimal_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0555da7c-cd7f-41ed-9d84-3656fb88babb",
   "metadata": {},
   "source": [
    "## Plot Optimal Butyrate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b2460bd-b62d-4a7f-b370-8da1ccf2c696",
   "metadata": {},
   "outputs": [],
   "source": [
    "optimal_plot = (ggplot(optimal_but_melt, aes(x = 'type', y = 'flux'))\n",
    "                    +geom_boxplot(aes(fill = 'type'))\n",
    "                    +scale_fill_manual(values = {'Optimal': '#b2d0b1',\n",
    "                                                 'Standard': '#fee3b2'\n",
    "                                                })\n",
    "                    +labs(y = 'Predicted Butyrate, mmol/gDW/h', x = '',title = 'Butyrate', fill = 'Probiotic Treatment')\n",
    "                    +theme_minimal()    \n",
    "                    +theme(figure_size=(10, 5),\n",
    "                           legend_position='none',\n",
    "                           axis_title=element_text(size=18),\n",
    "        axis_text=element_text(size=18),\n",
    "        plot_title=element_text(size=20))               )\n",
    "ggsave(optimal_plot, '../figures/optimal_butyrate.svg',dpi =300)\n",
    "optimal_plot\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfbf88b3-5d93-49e8-b866-c21772216e2e",
   "metadata": {},
   "source": [
    "## Reformat Indicies for Supplementary Figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d79ebf2a-c2e8-40e6-9914-3ab357892ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "production_ppa_pvt_noHF.index = production_ppa_pvt_noHF.index.str.replace('_+', ', Probiotic')\n",
    "production_ppa_pvt_noHF.index = production_ppa_pvt_noHF.index.str.replace('_-', ', No Probiotic')\n",
    "production_ppa_pvt_noHF = production_ppa_pvt_noHF.T\n",
    "\n",
    "production_but_pvt_noHF.index = production_but_pvt_noHF.index.str.replace('_+', ', Probiotic')\n",
    "production_but_pvt_noHF.index = production_but_pvt_noHF.index.str.replace('_-', ', No Probiotic')\n",
    "production_but_pvt_noHF = production_but_pvt_noHF.T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dd59dd0-8933-4f78-9dda-1d10a0455c48",
   "metadata": {},
   "source": [
    "## Plot Personalized Prediction Heatmaps "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8147cfe8-24af-4204-b999-28b5fb21a41b",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_indicies = production_but_pvt_noHF.idxmax(axis = 1)\n",
    "\n",
    "\n",
    "ax = sns.clustermap(production_but_pvt_noHF.T,\n",
    "                    cmap = 'Reds',\n",
    "                    figsize = (20,15),\n",
    "                    fmt='',\n",
    "                    xticklabels = False,\n",
    "                    col_colors = row_colors, \n",
    "                    col_cluster = False, \n",
    "                    row_cluster = False,\n",
    "                    cbar_pos= [0.05,0.05,.05,.72],\n",
    "                    annot_kws={'fontsize': 18, 'color':'white','verticalalignment': 'center'})\n",
    "\n",
    "for x in (max_indicies.index):\n",
    "    ax.ax_heatmap.add_patch(Rectangle((production_but_pvt_noHF.index.get_loc(x),\n",
    "                                      production_but_pvt_noHF.columns.get_loc(max_indicies[x])),\n",
    "                                       1, 1, fill=False, edgecolor='black', lw=3))\n",
    "\n",
    "\n",
    "ax.savefig(\"personalized_but.svg\", dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed10897f-b2d0-41e1-88c5-400abdd4004c",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_indicies = production_ppa_pvt_noHF.idxmax(axis = 1)\n",
    "\n",
    "\n",
    "ax = sns.clustermap(production_ppa_pvt_noHF.T,\n",
    "                    cmap = 'Reds',\n",
    "                    figsize = (20,15),\n",
    "                    fmt='',\n",
    "                    xticklabels = False,\n",
    "                    col_colors = row_colors, \n",
    "                    col_cluster = False, \n",
    "                    row_cluster = False,\n",
    "                    cbar_pos= [0.05,0.05,.05,.72],\n",
    "                    annot_kws={'fontsize': 18, 'color':'white','verticalalignment': 'center'})\n",
    "\n",
    "for x in (max_indicies.index):\n",
    "    ax.ax_heatmap.add_patch(Rectangle((production_ppa_pvt_noHF.index.get_loc(x),\n",
    "                                      production_ppa_pvt_noHF.columns.get_loc(max_indicies[x])),\n",
    "                                       1, 1, fill=False, edgecolor='black', lw=3))\n",
    "\n",
    "ax.savefig(\"personalized_ppa.svg\", dpi=300)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
