{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df2b5f6a-4075-4b9b-9d6c-33c254ca2ca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import micom\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sklearn.metrics\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from plotnine import *\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18c490ab-1379-41b0-a85b-7fd92c3b5af5",
   "metadata": {},
   "source": [
    "## Model Construction and Optimization\n",
    "This notebook constructs sample-specific MCMMs for samples in the WBF011 study, to compare predicted probiotic engraftment against experimental data. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd07717-8b9e-465b-8137-31c17085ea77",
   "metadata": {},
   "source": [
    "## Collect Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6e827fd7-b177-4a48-8665-469b06926075",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>time_point</th>\n",
       "      <th>treatment_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0223873</td>\n",
       "      <td>SS_65</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0253061</td>\n",
       "      <td>SS_60</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C0227725</td>\n",
       "      <td>SS_46</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0264833</td>\n",
       "      <td>SS_31</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C0252289</td>\n",
       "      <td>SS_67</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>C0228492</td>\n",
       "      <td>SS_118</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>C1064093</td>\n",
       "      <td>SS_102</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>C0231040</td>\n",
       "      <td>SS_76</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>C0263808</td>\n",
       "      <td>SS_71</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>C1010798</td>\n",
       "      <td>SS_124</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name subject_id   time_point treatment_group\n",
       "1   C0223873      SS_65  00_Baseline           wbf11\n",
       "2   C0253061      SS_60  00_Baseline           wbf11\n",
       "3   C0227725      SS_46  00_Baseline           wbf11\n",
       "4   C0264833      SS_31      12_Week         placebo\n",
       "5   C0252289      SS_67  00_Baseline           wbf11\n",
       "..       ...        ...          ...             ...\n",
       "78  C0228492     SS_118      12_Week         placebo\n",
       "79  C1064093     SS_102      12_Week           wbf11\n",
       "80  C0231040      SS_76      12_Week         placebo\n",
       "81  C0263808      SS_71      12_Week         placebo\n",
       "82  C1010798     SS_124      12_Week         placebo\n",
       "\n",
       "[82 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the Metadata Tabe\n",
    "metadata = pd.read_table('../data/Hiseq_metagenomic_202_190916 metadata_conditions.txt')\n",
    "metadata = metadata.drop(metadata.index[[0,83,84]])\n",
    "\n",
    "# Identify Subgroups\n",
    "baseline = metadata[metadata['time_point']=='00_Baseline']['Name']\n",
    "endpoint = metadata[metadata['time_point']=='12_Week']['Name']\n",
    "treatment = metadata[metadata['treatment_group']=='wbf11']['Name']\n",
    "placebo = metadata[metadata['treatment_group']=='placebo']['Name']\n",
    "subject_dict = metadata.set_index('Name')['subject_id'].to_dict()\n",
    "treatment_dict = metadata.set_index('Name')['treatment_group'].to_dict()\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0ce2d9-cf32-4905-8976-1c9fb29e968e",
   "metadata": {},
   "source": [
    "## Read Taxonomic Assignments from Kraken2\n",
    "Collect the species-level read counts, and separate samples into placebo and treatment samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "94b076e8-cef9-4982-8a3b-6b9c944d073b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_249096/3091618847.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/tmp/ipykernel_249096/3091618847.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/tmp/ipykernel_249096/3091618847.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/tmp/ipykernel_249096/3091618847.py:16: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "/tmp/ipykernel_249096/3091618847.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    }
   ],
   "source": [
    "# Load the Kraken taxonomic assignment data\n",
    "counts = pd.read_csv('../data/S_counts.csv', index_col=0)\n",
    "\n",
    "# Isolate Baseline Samples\n",
    "counts['sample'] = counts['sample'].str.split('_').str[0]\n",
    "counts_baseline = counts[counts['sample'].isin(baseline)]\n",
    "\n",
    "# Format for MICOM\n",
    "counts_baseline.rename(columns = {'sample':'sample_id', 's':'species'}, inplace = True)\n",
    "counts_baseline = counts_baseline.groupby(['sample_id','species']).sum(numeric_only = True).reset_index()\n",
    "counts_baseline['tot_reads'] = counts_baseline.groupby('sample_id')['reads'].transform('sum')\n",
    "counts_baseline['abundance'] = counts_baseline['reads']/counts_baseline['tot_reads']\n",
    "counts_baseline['id'] = counts_baseline['species'].str.replace(' ','_')\n",
    "\n",
    "# Isolate Treatment Group Samples\n",
    "counts_treatment = counts_baseline[counts_baseline['sample_id'].isin(treatment)]\n",
    "\n",
    "# Fix misannotation\n",
    "counts_treatment['id'] = counts_treatment['id'].str.replace('_(ex_Wegman_et_al._2014)','')\n",
    "counts_treatment['species'] = counts_treatment['species'].str.replace(' (ex Wegman et al. 2014)','')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab801192-cb9c-47d2-9ea8-c90862b4038c",
   "metadata": {},
   "source": [
    "## Add Probiotics to Treatment Arm\n",
    "Samples in the treatment arm are supplemented with a probiotic cocktail, wbf11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f516abc8-9947-4939-871d-4a54397b376c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>id</th>\n",
       "      <th>species</th>\n",
       "      <th>reads</th>\n",
       "      <th>tot_reads</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C0219269</td>\n",
       "      <td>Acidaminococcus_fermentans</td>\n",
       "      <td>Acidaminococcus fermentans</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3129594.0</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0219269</td>\n",
       "      <td>Acidaminococcus_intestini</td>\n",
       "      <td>Acidaminococcus intestini</td>\n",
       "      <td>409.0</td>\n",
       "      <td>3129594.0</td>\n",
       "      <td>0.000124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0219269</td>\n",
       "      <td>Actinomyces_naeslundii</td>\n",
       "      <td>Actinomyces naeslundii</td>\n",
       "      <td>38.0</td>\n",
       "      <td>3129594.0</td>\n",
       "      <td>0.000012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C0219269</td>\n",
       "      <td>Actinomyces_pacaensis</td>\n",
       "      <td>Actinomyces pacaensis</td>\n",
       "      <td>11.0</td>\n",
       "      <td>3129594.0</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0219269</td>\n",
       "      <td>Actinomyces_sp._oral_taxon_414</td>\n",
       "      <td>Actinomyces sp. oral taxon 414</td>\n",
       "      <td>21.0</td>\n",
       "      <td>3129594.0</td>\n",
       "      <td>0.000006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6402</th>\n",
       "      <td>C0265088</td>\n",
       "      <td>[Clostridium]_scindens</td>\n",
       "      <td>[Clostridium] scindens</td>\n",
       "      <td>28371.0</td>\n",
       "      <td>16964251.0</td>\n",
       "      <td>0.001589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6403</th>\n",
       "      <td>C0265088</td>\n",
       "      <td>[Ruminococcus]_gnavus</td>\n",
       "      <td>[Ruminococcus] gnavus</td>\n",
       "      <td>850972.0</td>\n",
       "      <td>16964251.0</td>\n",
       "      <td>0.047655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6404</th>\n",
       "      <td>C0265088</td>\n",
       "      <td>[Ruminococcus]_lactaris</td>\n",
       "      <td>[Ruminococcus] lactaris</td>\n",
       "      <td>6290.0</td>\n",
       "      <td>16964251.0</td>\n",
       "      <td>0.000352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6405</th>\n",
       "      <td>C0265088</td>\n",
       "      <td>[Ruminococcus]_torques</td>\n",
       "      <td>[Ruminococcus] torques</td>\n",
       "      <td>118751.0</td>\n",
       "      <td>16964251.0</td>\n",
       "      <td>0.006650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6406</th>\n",
       "      <td>C0265088</td>\n",
       "      <td>uncultured_Alistipes_sp.</td>\n",
       "      <td>uncultured Alistipes sp.</td>\n",
       "      <td>61.0</td>\n",
       "      <td>16964251.0</td>\n",
       "      <td>0.000003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6407 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id                              id  \\\n",
       "0     C0219269      Acidaminococcus_fermentans   \n",
       "1     C0219269       Acidaminococcus_intestini   \n",
       "2     C0219269          Actinomyces_naeslundii   \n",
       "3     C0219269           Actinomyces_pacaensis   \n",
       "4     C0219269  Actinomyces_sp._oral_taxon_414   \n",
       "...        ...                             ...   \n",
       "6402  C0265088          [Clostridium]_scindens   \n",
       "6403  C0265088           [Ruminococcus]_gnavus   \n",
       "6404  C0265088         [Ruminococcus]_lactaris   \n",
       "6405  C0265088          [Ruminococcus]_torques   \n",
       "6406  C0265088        uncultured_Alistipes_sp.   \n",
       "\n",
       "                             species     reads   tot_reads  abundance  \n",
       "0         Acidaminococcus fermentans      10.0   3129594.0   0.000003  \n",
       "1          Acidaminococcus intestini     409.0   3129594.0   0.000124  \n",
       "2             Actinomyces naeslundii      38.0   3129594.0   0.000012  \n",
       "3              Actinomyces pacaensis      11.0   3129594.0   0.000003  \n",
       "4     Actinomyces sp. oral taxon 414      21.0   3129594.0   0.000006  \n",
       "...                              ...       ...         ...        ...  \n",
       "6402          [Clostridium] scindens   28371.0  16964251.0   0.001589  \n",
       "6403           [Ruminococcus] gnavus  850972.0  16964251.0   0.047655  \n",
       "6404         [Ruminococcus] lactaris    6290.0  16964251.0   0.000352  \n",
       "6405          [Ruminococcus] torques  118751.0  16964251.0   0.006650  \n",
       "6406        uncultured Alistipes sp.      61.0  16964251.0   0.000003  \n",
       "\n",
       "[6407 rows x 6 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def add_all_probiotic(taxonomy):\n",
    "    \"\"\"Adds probiotic cocktail of AMUC, CBEI, CBUT, BINF, and EHAL to existing taxonomy table at 1% RA each\"\"\"\n",
    "    taxonomy_probiotic_total = pd.DataFrame()\n",
    "    for sample in taxonomy['sample_id'].unique():\n",
    "        taxonomy_reduced = taxonomy[taxonomy['sample_id'] == sample].copy()\n",
    "        taxonomy_reduced['abundance'] = taxonomy_reduced['abundance']*0.95\n",
    "        taxonomy_probiotic = pd.concat([taxonomy_reduced,pd.DataFrame({\n",
    "                  'sample_id':[sample, sample, sample, sample, sample], \n",
    "                  'species': ['Akkermansia muciniphila','Clostridium beijerinckii',\n",
    "                              'Clostridium butyricum','Bifidobacterium longum',\n",
    "                              'Anaerobutyricum hallii'],\n",
    "                  'abundance':[0.01, 0.01, 0.01, 0.01, 0.01], \n",
    "                  'id':['Akkermansia_muciniphila','Clostridium_beijerinckii',\n",
    "                              'Clostridium_butyricum','Bifidobacterium_longum',\n",
    "                              'Anaerobutyricum_hallii']})])\n",
    "        taxonomy_probiotic_total = pd.concat([taxonomy_probiotic_total, taxonomy_probiotic])\n",
    "\n",
    "    return taxonomy_probiotic_total\n",
    "# Add probiotics\n",
    "counts_probiotic = add_all_probiotic(counts_treatment)\n",
    "\n",
    "# Combine duplicate taxa\n",
    "counts_probiotic = counts_probiotic.groupby(['sample_id','id','species']).sum().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef94997c-5d3c-4617-88ab-ec509538f6c0",
   "metadata": {},
   "source": [
    "## Build Models\n",
    "Build MCMMs for samples in the treatment arm and the placebo arm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c820157d-b81e-41f4-8844-0df66bcc19b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "93111751c5ff4ab495a6fe71f556d4ef",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Output()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2;36m[06/13/25 02:40:57]\u001b[0m\u001b[2;36m \u001b[0m\u001b[33mWARNING \u001b[0m Less than \u001b[1;36m50\u001b[0m% of the abundance     \u001b]8;id=171813;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py\u001b\\\u001b[2mcommunity.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=35074;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py#229\u001b\\\u001b[2m229\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m                    \u001b[0m         could be matched to the model      \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         database. Model `C0223873` may not \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         be representative of the sample    \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m[06/13/25 02:40:57]\u001b[0m\u001b[2;36m \u001b[0m\u001b[33mWARNING \u001b[0m Less than \u001b[1;36m50\u001b[0m% of the abundance     \u001b]8;id=679142;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py\u001b\\\u001b[2mcommunity.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=342991;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py#229\u001b\\\u001b[2m229\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m                    \u001b[0m         could be matched to the model      \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         database. Model `C0224389` may not \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         be representative of the sample    \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m[06/13/25 02:40:57]\u001b[0m\u001b[2;36m \u001b[0m\u001b[33mWARNING \u001b[0m Less than \u001b[1;36m50\u001b[0m% of the abundance     \u001b]8;id=319364;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py\u001b\\\u001b[2mcommunity.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=23423;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py#229\u001b\\\u001b[2m229\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m                    \u001b[0m         could be matched to the model      \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         database. Model `C0232068` may not \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         be representative of the sample    \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m[06/13/25 02:49:03]\u001b[0m\u001b[2;36m \u001b[0m\u001b[33mWARNING \u001b[0m Less than \u001b[1;36m50\u001b[0m% of the abundance     \u001b]8;id=509366;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py\u001b\\\u001b[2mcommunity.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=725754;file:///users/nbohmann/miniconda3/envs/micom2024/lib/python3.9/site-packages/micom/community.py#229\u001b\\\u001b[2m229\u001b[0m\u001b]8;;\u001b\\\n",
      "\u001b[2;36m                    \u001b[0m         could be matched to the model      \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         database. Model `C0232589` may not \u001b[2m                \u001b[0m\n",
      "\u001b[2;36m                    \u001b[0m         be representative of the sample    \u001b[2m                \u001b[0m\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub message rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_msg_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_msg_rate_limit=1000.0 (msgs/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Build Models\n",
    "agora_augmented = '../agora_models/'\n",
    "\n",
    "manifest_agora1 = micom.workflows.build(counts_probiotic, \n",
    "                                 model_db=agora_augmented, \n",
    "                                 out_folder='../WBF011_models',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
