{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df2b5f6a-4075-4b9b-9d6c-33c254ca2ca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import micom\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sklearn.metrics\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from plotnine import *\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18c490ab-1379-41b0-a85b-7fd92c3b5af5",
   "metadata": {},
   "source": [
    "# Model Construction\n",
    "This notebook constructs sample-specific MCMMs for samples in the WBF011 study, to compare predicted probiotic engraftment against experimental data. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd07717-8b9e-465b-8137-31c17085ea77",
   "metadata": {},
   "source": [
    "## Collect Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e827fd7-b177-4a48-8665-469b06926075",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the Metadata Tabe\n",
    "metadata = pd.read_table('../data/Hiseq_metagenomic_202_190916 metadata_conditions.txt')\n",
    "metadata = metadata.drop(metadata.index[[0,83,84]])\n",
    "\n",
    "# Identify Subgroups\n",
    "baseline = metadata[metadata['time_point']=='00_Baseline']['Name']\n",
    "endpoint = metadata[metadata['time_point']=='12_Week']['Name']\n",
    "treatment = metadata[metadata['treatment_group']=='wbf11']['Name']\n",
    "placebo = metadata[metadata['treatment_group']=='placebo']['Name']\n",
    "subject_dict = metadata.set_index('Name')['subject_id'].to_dict()\n",
    "treatment_dict = metadata.set_index('Name')['treatment_group'].to_dict()\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0ce2d9-cf32-4905-8976-1c9fb29e968e",
   "metadata": {},
   "source": [
    "## Read Taxonomic Assignments from Kraken2\n",
    "Collect the species-level read counts, and separate samples into placebo and treatment samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94b076e8-cef9-4982-8a3b-6b9c944d073b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the Kraken taxonomic assignment data\n",
    "counts = pd.read_csv('../data/S_counts.csv', index_col=0)\n",
    "\n",
    "# Isolate Baseline Samples\n",
    "counts['sample'] = counts['sample'].str.split('_').str[0]\n",
    "counts_baseline = counts[counts['sample'].isin(baseline)]\n",
    "\n",
    "# Format for MICOM\n",
    "counts_baseline.rename(columns = {'sample':'sample_id', 's':'species'}, inplace = True)\n",
    "counts_baseline = counts_baseline.groupby(['sample_id','species']).sum(numeric_only = True).reset_index()\n",
    "counts_baseline['tot_reads'] = counts_baseline.groupby('sample_id')['reads'].transform('sum')\n",
    "counts_baseline['abundance'] = counts_baseline['reads']/counts_baseline['tot_reads']\n",
    "counts_baseline['id'] = counts_baseline['species'].str.replace(' ','_')\n",
    "counts_baseline['id'] = counts_baseline['id'].str.replace('_(ex_Wegman_et_al._2014)','')\n",
    "counts_baseline['species'] = counts_baseline['species'].str.replace(' (ex Wegman et al. 2014)','')\n",
    "\n",
    "# Isolate Treatment Group Samples\n",
    "counts_treatment = counts_baseline[counts_baseline['sample_id'].isin(treatment)]\n",
    "counts_placebo = counts_baseline[counts_baseline['sample_id'].isin(placebo)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab801192-cb9c-47d2-9ea8-c90862b4038c",
   "metadata": {},
   "source": [
    "## Add Probiotics to Treatment Arm\n",
    "Samples in the treatment arm are supplemented with a probiotic cocktail, wbf11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f516abc8-9947-4939-871d-4a54397b376c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_all_probiotic(taxonomy):\n",
    "    \"\"\"Adds probiotic cocktail of AMUC, CBEI, CBUT, BINF, and EHAL to existing taxonomy table at 1% RA each\"\"\"\n",
    "    taxonomy_probiotic_total = pd.DataFrame()\n",
    "    for sample in taxonomy['sample_id'].unique():\n",
    "        taxonomy_reduced = taxonomy[taxonomy['sample_id'] == sample].copy()\n",
    "        taxonomy_reduced['abundance'] = taxonomy_reduced['abundance']*0.95\n",
    "        taxonomy_probiotic = pd.concat([taxonomy_reduced,pd.DataFrame({\n",
    "                  'sample_id':[sample, sample, sample, sample, sample], \n",
    "                  'species': ['Akkermansia muciniphila','Clostridium beijerinckii',\n",
    "                              'Clostridium butyricum','Bifidobacterium longum',\n",
    "                              'Anaerobutyricum hallii'],\n",
    "                  'abundance':[0.01, 0.01, 0.01, 0.01, 0.01], \n",
    "                  'id':['Akkermansia_muciniphila','Clostridium_beijerinckii',\n",
    "                              'Clostridium_butyricum','Bifidobacterium_longum',\n",
    "                              'Anaerobutyricum_hallii']})])\n",
    "        taxonomy_probiotic_total = pd.concat([taxonomy_probiotic_total, taxonomy_probiotic])\n",
    "\n",
    "    return taxonomy_probiotic_total\n",
    "# Add probiotics\n",
    "counts_probiotic = add_all_probiotic(counts_treatment)\n",
    "\n",
    "# Combine duplicate taxa\n",
    "counts_probiotic = counts_probiotic.groupby(['sample_id','id','species']).sum().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef94997c-5d3c-4617-88ab-ec509538f6c0",
   "metadata": {},
   "source": [
    "## Build Models\n",
    "Build MCMMs for samples in the treatment arm and the placebo arm"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a3140cf2-6804-4e45-b67f-44dde03c8c10",
   "metadata": {},
   "source": [
    "# Build Models\n",
    "agora = '../agora_models/data'\n",
    "\n",
    "# Treatment group models\n",
    "micom.workflows.build(counts_probiotic, \n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../WBF011_treated',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n",
    "\n",
    "# Treatment group models w/o probiotics\n",
    "micom.workflows.build(counts_treatment, \n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../WBF011_noProbiotic',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n",
    "\n",
    "# Placebo group models\n",
    "micom.workflows.build(counts_placebo, \n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../Placebo',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
