{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "197cefdf-4b19-4546-8934-9c556af8e982",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from plotnine import *\n",
    "import scipy.stats\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc34367d-a08c-4b3c-a4f4-07d098c6e1f9",
   "metadata": {},
   "source": [
    "## Collect Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9c4f88c-0491-443d-a7ee-5df34b7f7459",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Collect metadata \n",
    "metadata = pd.read_table('../data/Hiseq_metagenomic_202_190916 metadata_conditions.txt')\n",
    "\n",
    "# Drop header and control rows\n",
    "metadata = metadata.drop(metadata.index[[0,83,84]])\n",
    "\n",
    "# Create dictionary with subject IDs\n",
    "subject_dict = metadata.set_index('Name')['subject_id'].to_dict()\n",
    "treatment_dict = metadata.set_index('Name')['treatment_group'].to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be8cc105-0d9d-4bdb-abc9-e555952a2b6c",
   "metadata": {},
   "source": [
    "## Import qPCR Table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc287df4-3ce9-474e-afb7-6bffce4fbabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Table\n",
    "qPCR = pd.read_table('../data/qPCR-reaction-table.tsv')\n",
    "\n",
    "# Rename EHAL to AHAL \n",
    "qPCR['Strain'] = qPCR['Strain'].str.replace('EHAL','AHAL')\n",
    "qPCR.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6320aff0-d2ec-460a-ab0c-953b32447c45",
   "metadata": {},
   "source": [
    "## Filter Table "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f07bd11d-e301-469a-9ee3-a4093d9af7a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove cycle thresholds of -inf\n",
    "qPCR_Ct = qPCR[qPCR['Ct']!= -np.inf].sort_values(by = 'Ct') \n",
    "\n",
    "# Remove reads that failed QC, filter\n",
    "qPCR_filter = qPCR_Ct[(qPCR_Ct['PassQC'] == True)&(qPCR_Ct['PassFilter'] == True)]\n",
    "\n",
    "# Remove less efficient primers for EHAL, CBEI\n",
    "qPCR_filter = qPCR_filter.loc[qPCR['Primers'] != 'EHAL_AN']\n",
    "qPCR_filter = qPCR_filter.loc[qPCR['Primers'] != 'CBEI_AB']\n",
    "\n",
    "# Format column types\n",
    "qPCR_filter['Week'] = qPCR_filter['Week'].astype('str')\n",
    "\n",
    "# Sort table\n",
    "qPCR_filter = qPCR_filter.sort_values(by = ['Strain','Week'])\n",
    "\n",
    "qPCR_filter.Strain.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74c6e8a2-6955-4b44-872b-f1d941741ec1",
   "metadata": {},
   "source": [
    "## Calculate Inverse Cycle Threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b15b0053-7ff2-4e92-bf6f-d749c9e3327b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate inverse\n",
    "qPCR_filter['Ct_inv'] = 1/qPCR_filter['Ct']\n",
    "qPCR_filter = qPCR_filter[qPCR_filter['Week']!= '16']\n",
    "qPCR_filter = qPCR_filter[qPCR_filter['Treatment'] != 'WBF-010']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38304411-b7f3-46ee-a83e-a3b3e7d1467e",
   "metadata": {},
   "source": [
    "## Plot qPCR Data \n",
    "Plot 1/Ct across study arms for each strain and week"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42747210-0239-416b-91bf-11f76d537345",
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# Set all styling consistently BEFORE plotting\n",
    "sns.set_theme(style=\"whitegrid\")       # Ensures white grid\n",
    "sns.set_context(\"talk\", font_scale=1.2)  # Sets all font sizes reasonably large\n",
    "sns.set_palette(\"colorblind\")          # Applies colorblind-safe palette\n",
    "\n",
    "# Now create your plots\n",
    "treatments = np.array(['Placebo', 'WBF-011'], dtype='object')\n",
    "fig, axes = plt.subplots(1, 2, figsize=(16, 9), sharey=True)\n",
    "\n",
    "# Loop over treatments and make plots\n",
    "for i, treatment in enumerate(treatments):\n",
    "    ax = axes[i]\n",
    "\n",
    "    treatment_data = qPCR_filter[qPCR_filter[\"Treatment\"] == treatment]\n",
    "\n",
    "    # Calculate strain thresholds\n",
    "    strain_thresholds = treatment_data.loc[(treatment_data['Week'] == '0')].groupby('Strain')['Ct_inv'].mean().to_dict()\n",
    "\n",
    "    # Manually input strain threshold for BINF to match WBF-011 baseline (no hits in placebo)\n",
    "    if len(strain_thresholds) < 5:\n",
    "        strain_thresholds['BINF'] = 0.028\n",
    "\n",
    "    # Map Strain Thresholds\n",
    "    treatment_data['threshold'] = treatment_data['Strain'].map(strain_thresholds)\n",
    "\n",
    "    treatment_data = treatment_data.groupby(['Treatment','Strain','Week','Subject']).mean(numeric_only=True).reset_index()\n",
    "    treatment_data['Week'] = treatment_data['Week'].astype(int)\n",
    "    treatment_data = treatment_data.sort_values(by=['Strain','Week'])\n",
    "    treatment_data['Week'] = treatment_data['Week'].astype(str)\n",
    "\n",
    "    # Plot\n",
    "    sns.violinplot(data=treatment_data, x=\"Strain\", y=\"Ct_inv\", hue=\"Week\", \n",
    "                   split=False, alpha=0.5, inner=None, ax=ax, cut=0.01)\n",
    "    sns.stripplot(data=treatment_data, x=\"Strain\", y=\"Ct_inv\", hue=\"Week\", \n",
    "                  dodge=True, size=5, jitter=0.4, ax=ax, legend=False)\n",
    "\n",
    "    ax.set(ylabel='')\n",
    "    ax.set_title(f\"Detection Over Time - {treatment}\")\n",
    "    # Add threshold lines \n",
    "    x_positions = {strain: idx for idx, strain in enumerate(treatment_data[\"Strain\"].unique())}\n",
    "    for strain in x_positions:\n",
    "        x_pos = x_positions[strain]\n",
    "        ax.hlines(y=strain_thresholds[strain], xmin=x_pos - 0.5, xmax=x_pos + 0.5, \n",
    "                  color='red', linestyle='--', linewidth=1.5)\n",
    "\n",
    "fig.supylabel('1/Ct Value')\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"../figures/qPCR_threshold.svg\", dpi=300, bbox_inches=\"tight\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "604dd25f-a470-41e6-a9fc-9ed9ea01f215",
   "metadata": {},
   "source": [
    "## Calculate Signficance For Treatment Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34f9617d-c2a2-49d8-bba8-ef01fa107e6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats\n",
    "\n",
    "\n",
    "for strain in qPCR_filter['Strain'].unique():\n",
    "    res = scipy.stats.mannwhitneyu(treatment_data[(treatment_data['Strain'] == strain)&(treatment_data['Week']==0)]['Ct_inv'],\n",
    "                       treatment_data[(treatment_data['Strain'] == strain)&(treatment_data['Week']==12)]['Ct_inv'])\n",
    "    print(strain+': '+str(res[1]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
