{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "03495f08-0f9d-49f1-a635-643d2b539781",
   "metadata": {},
   "source": [
    "# Validation of Probiotic Growth Predictions\n",
    "This notebook simulates the growth of MCMMs and validates the predicted probiotic growth rates for participants in Validation Study B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e5cdb39-913c-478a-9f79-22642360ff35",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import micom\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from plotnine import *\n",
    "import scipy.stats\n",
    "from sklearn.metrics import cohen_kappa_score\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1383651-5c98-4067-aae8-6312b71023f2",
   "metadata": {},
   "source": [
    "## Grab Metadata\n",
    "Get the metadata provided on the NCBI SRA so we can identify samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a767dd02-0eac-4644-9ac2-9f32c9cdf1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Metadata \n",
    "metadata = pd.read_csv('../data/studyB_metadata.csv') \n",
    "\n",
    "# Split identifiers into separate columns\n",
    "metadata[['cohort','subject_id']] = metadata['Cohort;_Subject_ID'].str.split(';', expand = True) \n",
    "\n",
    "# Isolate columns of interest\n",
    "metadata_df = metadata[['Run','study_timepoint','cohort', 'subject_id']].rename(columns = {'Run':'sample_id'}) \n",
    "\n",
    "# Create subject dictionary\n",
    "subject_dict = metadata_df.set_index('sample_id')['subject_id'].to_dict()\n",
    "\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95abf4b6-cf37-49d8-ab53-49c7d1ed31af",
   "metadata": {},
   "source": [
    "# Load Sequencing Data\n",
    "Load the species-level count Kraken2 output from MEDI, and merge with metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6144f9d1-d062-4661-b624-5421e4b62ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load sequencing data\n",
    "s_counts = pd.read_csv('../data/studyB_S_counts.csv') \n",
    "\n",
    "# Match sample ID format to metadata\n",
    "s_counts['sample_id'] = s_counts['sample_id'].str.replace(r'^S_', '', regex=True) \n",
    "\n",
    "# Merge with metadata \n",
    "s_counts = s_counts.merge(metadata_df, on='sample_id', how='left')\n",
    "\n",
    "# Sort table\n",
    "s_counts.sort_values(by = ['cohort','study_timepoint'], inplace = True)\n",
    "s_counts.subject_id.nunique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dfd9d3b-cd09-446f-9d15-471678762f6e",
   "metadata": {},
   "source": [
    "## Isolate Metagenomic Endpoint Measurements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3879d335-1de6-41db-ac0d-6d9a4ee378f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Isolate probiotic species \n",
    "counts_endpoint = s_counts[s_counts['name'].str.contains('|'.join(['Enterocloster bolteae', 'Anaerotruncus colihominis',\n",
    "                                                                      'Sellimonas intestinalis', 'symbiosum', \n",
    "                                                                      'Dorea longicatena','Flavonifractor plautii']))]\n",
    "\n",
    "# Identify cohorts of interest \n",
    "counts_endpoint = counts_endpoint[counts_endpoint['cohort'].str.contains('|'.join(['1','2','3','4','5']))]\n",
    "\n",
    "# Rename columns\n",
    "counts_endpoint.rename(columns = {'name':'species','fraction_total_reads':'abundance'}, inplace = True)\n",
    "\n",
    "# Filter to timepoints of interest \n",
    "counts_endpoint = counts_endpoint[(counts_endpoint['study_timepoint'] == 'Week 12')|\n",
    "    (counts_endpoint['study_timepoint'] == 'Day 0')][['species','abundance','study_timepoint','subject_id','cohort']]\n",
    "\n",
    "# Pivot data \n",
    "counts_endpoint = pd.pivot_table(counts_endpoint, \n",
    "                                 index = ['subject_id','species', 'cohort'],\n",
    "                                 columns = 'study_timepoint',\n",
    "                                 values = 'abundance').fillna(0.0).reset_index()\n",
    "\n",
    "# Drop sample missing follow-up\n",
    "counts_endpoint = counts_endpoint[counts_endpoint['subject_id'] != ' Subject 102']\n",
    "                                \n",
    "# Identify threshold for engraftment \n",
    "threshold_ratio = 1.5\n",
    "\n",
    "# Binarize engraftment by increase over threshold\n",
    "counts_endpoint.loc[threshold_ratio*counts_endpoint['Day 0']>=counts_endpoint['Week 12'], 'engraftment'] = 0\n",
    "counts_endpoint.loc[threshold_ratio*counts_endpoint['Day 0']<counts_endpoint['Week 12'], 'engraftment'] = 1\n",
    "\n",
    "# No engraftment for abundance below 0.5%\n",
    "counts_endpoint.loc[counts_endpoint['Week 12']<0.005, 'engraftment'] = 0\n",
    "\n",
    "# Fix formating for C. symbiosum\n",
    "counts_endpoint['species'] = counts_endpoint['species'].str.replace('[Clostridium]','Clostridium')\n",
    "\n",
    "counts_endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b030c03-4deb-4e11-910e-42aec4ff04b5",
   "metadata": {},
   "source": [
    "## Prepare MCMM Medium\n",
    "Load standard european medium as assumed background diet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86cd0624-2e70-4ba0-8f22-ff474681bc22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Medium \n",
    "eu_medium = pd.read_csv('../european_medium.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3878df1-7bd7-4f99-9ddc-424180dfa255",
   "metadata": {},
   "source": [
    "## Grow Models\n",
    "Simulate model growth with high tradeoff parameter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14ab89a6-8904-4480-bd3f-e011fc41600f",
   "metadata": {},
   "outputs": [],
   "source": [
    "manifest_probiotic = pd.read_csv('../VE303_treated/manifest.csv')\n",
    "\n",
    "\n",
    "growth = micom.workflows.grow(manifest_probiotic, \n",
    "                                  model_folder = '../VE303_treated/',\n",
    "                                  medium = eu_medium,\n",
    "                                  tradeoff = 0.99, \n",
    "                                  strategy = 'none',\n",
    "                                  threads = 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7332119b-c761-4135-97c8-51967f351a8b",
   "metadata": {},
   "source": [
    "## Collect Predicted Growth Rates\n",
    "Isolate predictions for probiotic species a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "445c3099-7b2c-4509-8692-d2fdafe8d621",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gather growth rates\n",
    "rates_probiotic = growth.growth_rates\n",
    "\n",
    "# Isolate probiotics\n",
    "\n",
    "rates_probiotic = rates_probiotic[rates_probiotic.index.isin(['Enterocloster_bolteae', \n",
    "                                                              'Anaerotruncus_colihominis',\n",
    "                                                              'Sellimonas_intestinalis', \n",
    "                                                              'Dorea_longicatena', \n",
    "                                                              'Clostridium_symbiosum', \n",
    "                                                              'Flavonifractor_plautii',\n",
    "                                                             ])].reset_index()\n",
    "# Log scale growth rates\n",
    "rates_probiotic['growth_rate_scaled'] = np.log10(rates_probiotic[['growth_rate']])\n",
    "\n",
    "# Threshold log-scaled growth rates\n",
    "rates_probiotic.loc[rates_probiotic['growth_rate_scaled']>=-2, 'MCMM_engraftment'] = 1\n",
    "rates_probiotic.loc[rates_probiotic['growth_rate_scaled']<-2, 'MCMM_engraftment'] = 0\n",
    "\n",
    "# Reformat Names \n",
    "rates_probiotic['compartments'] = rates_probiotic['compartments'].str.replace('_Clostridium_', 'Clostridium')\n",
    "rates_probiotic['compartments'] = rates_probiotic['compartments'].str.replace('_',' ')\n",
    "\n",
    "rates_probiotic"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c9bc6f7-ebfd-4e87-93fb-a0c7d8194fd7",
   "metadata": {},
   "source": [
    "## Create Agreement Heatmap\n",
    "Find agreement between predictions and measures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "572c708a-8ce5-4512-b948-40eb456a7848",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Metagenome Engraftment Pivot Table\n",
    "df1 = pd.pivot_table(counts_endpoint, \n",
    "                    index='species',\n",
    "                    columns='subject_id',\n",
    "                    values='engraftment').fillna(0.0)\n",
    "\n",
    "# Create MCMM Engraftment Pivot Table\n",
    "df2 = pd.pivot_table(rates_probiotic, \n",
    "                    index='compartments',\n",
    "                    columns='sample_id',\n",
    "                    values='MCMM_engraftment').fillna(0.0)\n",
    "\n",
    "# Ensure both dataframes are numeric (convert from strings if necessary)\n",
    "df1 = df1.astype(float)\n",
    "df2 = df2.astype(float)\n",
    "\n",
    "# Calculate agreement matrix\n",
    "agreement_matrix = (df1 == df2).astype(int)\n",
    "agreement_matrix.index.name = 'Probiotic Species'\n",
    "agreement_matrix.columns.name = 'Samples'\n",
    "agreement_matrix.index = agreement_matrix.index.str.replace('_', ' ')\n",
    "# Plot heatmap of agreement\n",
    "plt.figure(figsize=(18, 12))\n",
    "sns.set(font_scale=4) \n",
    "ax = sns.heatmap(agreement_matrix, cmap='coolwarm_r',\n",
    "                 annot=False, cbar=False, square=False, \n",
    "                 linewidths=1, linecolor='white')\n",
    "ax.set_xticklabels([])\n",
    "plt.title('Agreement Between MCMM and Metagenomes')\n",
    "plt.savefig('../figures/agreement_heatmap_studyB.svg', dpi=300, bbox_inches='tight')  # Save as PNG with 300 DPI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18ff41eb-b4cf-4418-84fd-ee26ba294544",
   "metadata": {},
   "source": [
    "## Calculate Cohen's Kappa\n",
    "Measure the agreement "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c967a4a-9e60-46e1-8c4e-2890da6c9ac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Flatten predictions and measures\n",
    "ratings1 = df1.values.flatten()\n",
    "ratings2 = df2.values.flatten()\n",
    "\n",
    "# Compute Cohen's kappa\n",
    "kappa = cohen_kappa_score(ratings1, ratings2)\n",
    "\n",
    "print(f\"Cohen's kappa: {kappa:.4f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce7a1ba2-eeec-42cc-9f5b-e2df937dd95d",
   "metadata": {},
   "source": [
    "## Calculate % Agreement\n",
    "Calculate agreement as a percentage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c796d06a-5fd9-4b6b-9e05-9417f94dd01b",
   "metadata": {},
   "outputs": [],
   "source": [
    "agree_ones = agreement_matrix.sum().sum()\n",
    "growth_ones = df2.sum().sum()\n",
    "total = agreement_matrix.size\n",
    "print(str((agree_ones/total).round(4)*100)+'%')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0015b41d-cbbe-4e19-9619-e8c42f51872d",
   "metadata": {},
   "source": [
    "## Plot Confusion Matricies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d890cd9e-7d31-4d20-aece-73fe58842c9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "strains = df1.index\n",
    "\n",
    "FONTS = {\n",
    "    \"suptitle\": 32,\n",
    "    \"title\": 28,\n",
    "    \"axis_labels\": 28,\n",
    "    \"tick_labels\": 25,\n",
    "    \"heatmap_annotations\": 28\n",
    "}\n",
    "\n",
    "sns.set(style=\"whitegrid\", font_scale=FONTS[\"tick_labels\"]/10)  # scale factor for seaborn\n",
    "# Set up subplot grid\n",
    "fig, axes = plt.subplots(3,3, figsize=(18, 18))\n",
    "axes = axes.flatten()  # Flatten for easy iteration\n",
    "\n",
    "\n",
    "\n",
    "# Define consistent color limits\n",
    "vmin, vmax = 0, 20\n",
    "\n",
    "# Aggregate all results\n",
    "all_y_true = df1.values.flatten()  # Flatten across all strains\n",
    "all_y_pred = df2.values.flatten()\n",
    "\n",
    "# Compute overall confusion matrix\n",
    "cm_agg = confusion_matrix(all_y_true, all_y_pred)\n",
    "\n",
    "# Ensure correct shape for formatting\n",
    "if cm_agg.shape == (1, 1):\n",
    "    cm_agg = np.array([[cm_agg[0, 0], 0], [0, 0]])\n",
    "\n",
    "# Plot aggregated confusion matrix\n",
    "sns.heatmap(cm_agg, annot=True, fmt=\"d\", cmap=\"Reds\",\n",
    "            xticklabels=[\"No Growth\", \"Growth\"],\n",
    "            yticklabels=[\"No Growth\", \"Growth\"],\n",
    "            ax=axes[0],\n",
    "            vmin=vmin*6, vmax=vmax*6,\n",
    "            annot_kws={\"fontsize\": FONTS[\"heatmap_annotations\"]})\n",
    "\n",
    "ax = axes[0]\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=0)\n",
    "ax.set_yticklabels(ax.get_yticklabels(), rotation=90)\n",
    "# Loop through each strain\n",
    "for i, strain in enumerate(strains):\n",
    "    y_true = df1.loc[strain].values.flatten()\n",
    "    y_pred = df2.loc[strain].values.flatten()\n",
    "\n",
    "    cm = confusion_matrix(y_true, y_pred)\n",
    "\n",
    "    if cm.shape == (1, 1):\n",
    "        cm = np.array([[cm[0, 0], 0], [0, 0]])\n",
    "\n",
    "    sns.heatmap(cm, annot=True, fmt=\"d\", cmap=\"Blues\",\n",
    "                xticklabels=[\"No Growth\", \"Growth\"],\n",
    "                yticklabels=[\"No Growth\", \"Growth\"],\n",
    "                ax=axes[i+1],\n",
    "                vmin=vmin, vmax=vmax,\n",
    "                annot_kws={\"fontsize\": FONTS[\"heatmap_annotations\"]})\n",
    "    ax = axes[i+1]\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=0)\n",
    "    ax.set_yticklabels(ax.get_yticklabels(), rotation=90)\n",
    "fig.delaxes(axes[-1])\n",
    "fig.delaxes(axes[-2])\n",
    "\n",
    "axes[0].set_title(\"Aggregate \\n Across Strains\", fontsize=FONTS[\"title\"])\n",
    "axes[1].set_title('Anaerotruncus \\n colihominis', fontsize=FONTS[\"title\"])\n",
    "axes[2].set_title('Dorea \\n longicatena', fontsize=FONTS[\"title\"])\n",
    "axes[3].set_title('Enterocloster \\n bolteae', fontsize=FONTS[\"title\"])\n",
    "axes[4].set_title('Flavonifractor \\n plautii', fontsize=FONTS[\"title\"])\n",
    "axes[5].set_title('Sellimonas \\n intestinalis', fontsize=FONTS[\"title\"])\n",
    "axes[6].set_title('Clostridium \\n symbiosum', fontsize=FONTS[\"title\"])\n",
    "fig.suptitle(\"Confusion Matrices of Model Predictions vs. Metagenomes\", fontsize=FONTS[\"suptitle\"])\n",
    "fig.text(-0.02, 0.5, \"Model Prediction\", va='center', rotation='vertical', fontsize=FONTS[\"axis_labels\"])\n",
    "fig.text(0.5, -0.02, \"Metagenomic Ground Truth\", ha='center', fontsize=FONTS[\"axis_labels\"])\n",
    "\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig('../confusion_matricies_studyB.svg', dpi=300, bbox_inches='tight')  # Save as PNG with 300 DPI\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94f7e06e-8029-4c79-ae94-cb017a5eb895",
   "metadata": {},
   "source": [
    "## Calculate Fisher's Exact Test for Aggregate Result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24e13d1f-6a42-49ab-b8f7-bf53a6d57237",
   "metadata": {},
   "outputs": [],
   "source": [
    "scipy.stats.fisher_exact(cm_agg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a630a0f-89e4-468b-b1ef-f6bb2f28d409",
   "metadata": {},
   "source": [
    "## Sensitivity Analysis\n",
    "Test how growth rates thresholds impact prediction accuracy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffe3ca17-e648-4b90-b9a0-87f59242542f",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.DataFrame()\n",
    "for threshold in np.logspace(-5, 0.7, num=50):\n",
    "        # Threshold log-scaled growth rates\n",
    "        rates_probiotic.loc[rates_probiotic['growth_rate']>=threshold, 'MCMM_engraftment'] = 1\n",
    "        rates_probiotic.loc[rates_probiotic['growth_rate']<threshold, 'MCMM_engraftment'] = 0\n",
    "        \n",
    "        \n",
    "        # Create MCMM Engraftment Pivot Table\n",
    "        df2 = pd.pivot_table(rates_probiotic, \n",
    "                            index='compartments',\n",
    "                            columns='sample_id',\n",
    "                            values='MCMM_engraftment').fillna(0.0)\n",
    "        df2.index = df2.index.str.replace('_Clostridium_','Clostridium')\n",
    "        # Ensure both dataframes are numeric (convert from strings if necessary)\n",
    "        df2 = df2.astype(float)\n",
    "        df2 = df2[df1.columns]\n",
    "    \n",
    "        # Calculate agreement matrix\n",
    "    \n",
    "        agreement_matrix = (df1 == df2).astype(int)\n",
    "        \n",
    "        ratings1 = df1.values.flatten()\n",
    "        ratings2 = df2.values.flatten()\n",
    "        \n",
    "        # Compute Cohen's kappa\n",
    "        kappa = cohen_kappa_score(ratings1, ratings2)\n",
    "        \n",
    "        agree_ones = agreement_matrix.sum().sum()\n",
    "        growth_ones = df2.sum().sum()\n",
    "        total = agreement_matrix.size\n",
    "        \n",
    "        cm = confusion_matrix(ratings1, ratings2)\n",
    "        \n",
    "        if cm.shape == (1, 1):\n",
    "            cm = np.array([[cm[0, 0], 0], [0, 0]])\n",
    "        fisher = scipy.stats.fisher_exact(cm)\n",
    "        res = pd.concat([res, pd.DataFrame({'agreement':[(agree_ones/total)],\n",
    "                                            'kappa':[kappa],\n",
    "                                            'fisher':[fisher[1]],\n",
    "                                            'threshold':[threshold]\n",
    "                                           }\n",
    "                                          )\n",
    "                        ]\n",
    "                       )\n",
    "# Convert results into long format \n",
    "res_long = res.melt(id_vars='threshold', value_vars=['agreement', 'kappa'],\n",
    "                    var_name='metric', value_name='value')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d85f740-5aaa-4331-b0a9-032cbe8397c6",
   "metadata": {},
   "source": [
    "## Plot Sensitivity "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cf54a8b-2e7c-445a-a071-d3656e96d447",
   "metadata": {},
   "outputs": [],
   "source": [
    "sensitivity_plot = (ggplot(res_long, aes(x='threshold', y='value', color='metric'))\n",
    "                    + geom_line(size=1.2)\n",
    "                    + scale_x_log10()\n",
    "                    +geom_vline(xintercept = 0.01)\n",
    "                    + labs(y='Value', x='Growth Rate Threshold', color='Metric')\n",
    "                    +scale_color_discrete(limits = ['agreement','kappa'], labels = ['Percent Agreement','Cohens Kappa'])\n",
    "                    + theme_minimal()\n",
    "                   )\n",
    "\n",
    "ggsave(sensitivity_plot, '../figures/sensitivity_studyB.svg',dpi = 200)\n",
    "sensitivity_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "193f8aae-573c-4918-9782-943196560071",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
