{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df2b5f6a-4075-4b9b-9d6c-33c254ca2ca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import micom\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sklearn.metrics\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from plotnine import *\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18c490ab-1379-41b0-a85b-7fd92c3b5af5",
   "metadata": {},
   "source": [
    "# Model Construction\n",
    "This notebook constructs sample-specific MCMMs for samples in the WBF011 study, to compare predicted probiotic engraftment against experimental data. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd07717-8b9e-465b-8137-31c17085ea77",
   "metadata": {},
   "source": [
    "## Collect Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6e827fd7-b177-4a48-8665-469b06926075",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>time_point</th>\n",
       "      <th>treatment_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C0223873</td>\n",
       "      <td>SS_65</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C0253061</td>\n",
       "      <td>SS_60</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C0227725</td>\n",
       "      <td>SS_46</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C0264833</td>\n",
       "      <td>SS_31</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>C0252289</td>\n",
       "      <td>SS_67</td>\n",
       "      <td>00_Baseline</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>C0228492</td>\n",
       "      <td>SS_118</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>C1064093</td>\n",
       "      <td>SS_102</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>wbf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>C0231040</td>\n",
       "      <td>SS_76</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>C0263808</td>\n",
       "      <td>SS_71</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>C1010798</td>\n",
       "      <td>SS_124</td>\n",
       "      <td>12_Week</td>\n",
       "      <td>placebo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Name subject_id   time_point treatment_group\n",
       "1   C0223873      SS_65  00_Baseline           wbf11\n",
       "2   C0253061      SS_60  00_Baseline           wbf11\n",
       "3   C0227725      SS_46  00_Baseline           wbf11\n",
       "4   C0264833      SS_31      12_Week         placebo\n",
       "5   C0252289      SS_67  00_Baseline           wbf11\n",
       "..       ...        ...          ...             ...\n",
       "78  C0228492     SS_118      12_Week         placebo\n",
       "79  C1064093     SS_102      12_Week           wbf11\n",
       "80  C0231040      SS_76      12_Week         placebo\n",
       "81  C0263808      SS_71      12_Week         placebo\n",
       "82  C1010798     SS_124      12_Week         placebo\n",
       "\n",
       "[82 rows x 4 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the Metadata Tabe\n",
    "metadata = pd.read_table('../data/Hiseq_metagenomic_202_190916 metadata_conditions.txt')\n",
    "metadata = metadata.drop(metadata.index[[0,83,84]])\n",
    "\n",
    "# Identify Subgroups\n",
    "baseline = metadata[metadata['time_point']=='00_Baseline']['Name']\n",
    "endpoint = metadata[metadata['time_point']=='12_Week']['Name']\n",
    "treatment = metadata[metadata['treatment_group']=='wbf11']['Name']\n",
    "placebo = metadata[metadata['treatment_group']=='placebo']['Name']\n",
    "subject_dict = metadata.set_index('Name')['subject_id'].to_dict()\n",
    "treatment_dict = metadata.set_index('Name')['treatment_group'].to_dict()\n",
    "\n",
    "metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0ce2d9-cf32-4905-8976-1c9fb29e968e",
   "metadata": {},
   "source": [
    "## Read Taxonomic Assignments from Kraken2\n",
    "Collect the species-level read counts, and separate samples into placebo and treatment samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "94b076e8-cef9-4982-8a3b-6b9c944d073b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_267718/921169281.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>species</th>\n",
       "      <th>reads</th>\n",
       "      <th>tot_reads</th>\n",
       "      <th>abundance</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>C0219524</td>\n",
       "      <td>Acidaminococcus fermentans</td>\n",
       "      <td>123</td>\n",
       "      <td>5397417</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>Acidaminococcus_fermentans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>C0219524</td>\n",
       "      <td>Acidaminococcus intestini</td>\n",
       "      <td>369</td>\n",
       "      <td>5397417</td>\n",
       "      <td>0.000068</td>\n",
       "      <td>Acidaminococcus_intestini</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>C0219524</td>\n",
       "      <td>Actinomyces naeslundii</td>\n",
       "      <td>76</td>\n",
       "      <td>5397417</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>Actinomyces_naeslundii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>C0219524</td>\n",
       "      <td>Actinomyces oris</td>\n",
       "      <td>110</td>\n",
       "      <td>5397417</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>Actinomyces_oris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>C0219524</td>\n",
       "      <td>Actinomyces pacaensis</td>\n",
       "      <td>13</td>\n",
       "      <td>5397417</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>Actinomyces_pacaensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11554</th>\n",
       "      <td>C0263817</td>\n",
       "      <td>[Clostridium] innocuum</td>\n",
       "      <td>37168</td>\n",
       "      <td>11089351</td>\n",
       "      <td>0.003352</td>\n",
       "      <td>[Clostridium]_innocuum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11555</th>\n",
       "      <td>C0263817</td>\n",
       "      <td>[Clostridium] scindens</td>\n",
       "      <td>108156</td>\n",
       "      <td>11089351</td>\n",
       "      <td>0.009753</td>\n",
       "      <td>[Clostridium]_scindens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11556</th>\n",
       "      <td>C0263817</td>\n",
       "      <td>[Ruminococcus] gnavus</td>\n",
       "      <td>368955</td>\n",
       "      <td>11089351</td>\n",
       "      <td>0.033271</td>\n",
       "      <td>[Ruminococcus]_gnavus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11557</th>\n",
       "      <td>C0263817</td>\n",
       "      <td>[Ruminococcus] lactaris</td>\n",
       "      <td>935</td>\n",
       "      <td>11089351</td>\n",
       "      <td>0.000084</td>\n",
       "      <td>[Ruminococcus]_lactaris</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11558</th>\n",
       "      <td>C0263817</td>\n",
       "      <td>[Ruminococcus] torques</td>\n",
       "      <td>4646</td>\n",
       "      <td>11089351</td>\n",
       "      <td>0.000419</td>\n",
       "      <td>[Ruminococcus]_torques</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6059 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id                     species   reads  tot_reads  abundance  \\\n",
       "290    C0219524  Acidaminococcus fermentans     123    5397417   0.000023   \n",
       "291    C0219524   Acidaminococcus intestini     369    5397417   0.000068   \n",
       "292    C0219524      Actinomyces naeslundii      76    5397417   0.000014   \n",
       "293    C0219524            Actinomyces oris     110    5397417   0.000020   \n",
       "294    C0219524       Actinomyces pacaensis      13    5397417   0.000002   \n",
       "...         ...                         ...     ...        ...        ...   \n",
       "11554  C0263817      [Clostridium] innocuum   37168   11089351   0.003352   \n",
       "11555  C0263817      [Clostridium] scindens  108156   11089351   0.009753   \n",
       "11556  C0263817       [Ruminococcus] gnavus  368955   11089351   0.033271   \n",
       "11557  C0263817     [Ruminococcus] lactaris     935   11089351   0.000084   \n",
       "11558  C0263817      [Ruminococcus] torques    4646   11089351   0.000419   \n",
       "\n",
       "                               id  \n",
       "290    Acidaminococcus_fermentans  \n",
       "291     Acidaminococcus_intestini  \n",
       "292        Actinomyces_naeslundii  \n",
       "293              Actinomyces_oris  \n",
       "294         Actinomyces_pacaensis  \n",
       "...                           ...  \n",
       "11554      [Clostridium]_innocuum  \n",
       "11555      [Clostridium]_scindens  \n",
       "11556       [Ruminococcus]_gnavus  \n",
       "11557     [Ruminococcus]_lactaris  \n",
       "11558      [Ruminococcus]_torques  \n",
       "\n",
       "[6059 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the Kraken taxonomic assignment data\n",
    "counts = pd.read_csv('../data/S_counts.csv', index_col=0)\n",
    "\n",
    "# Isolate Baseline Samples\n",
    "counts['sample'] = counts['sample'].str.split('_').str[0]\n",
    "counts_baseline = counts[counts['sample'].isin(baseline)]\n",
    "\n",
    "# Format for MICOM\n",
    "counts_baseline.rename(columns = {'sample':'sample_id', 's':'species'}, inplace = True)\n",
    "counts_baseline = counts_baseline.groupby(['sample_id','species']).sum(numeric_only = True).reset_index()\n",
    "counts_baseline['tot_reads'] = counts_baseline.groupby('sample_id')['reads'].transform('sum')\n",
    "counts_baseline['abundance'] = counts_baseline['reads']/counts_baseline['tot_reads']\n",
    "counts_baseline['id'] = counts_baseline['species'].str.replace(' ','_')\n",
    "counts_baseline['id'] = counts_baseline['id'].str.replace('_(ex_Wegman_et_al._2014)','')\n",
    "counts_baseline['species'] = counts_baseline['species'].str.replace(' (ex Wegman et al. 2014)','')\n",
    "\n",
    "# Isolate Treatment Group Samples\n",
    "counts_treatment = counts_baseline[counts_baseline['sample_id'].isin(treatment)]\n",
    "counts_placebo = counts_baseline[counts_baseline['sample_id'].isin(placebo)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab801192-cb9c-47d2-9ea8-c90862b4038c",
   "metadata": {},
   "source": [
    "## Add Probiotics to Treatment Arm\n",
    "Samples in the treatment arm are supplemented with a probiotic cocktail, wbf11"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f516abc8-9947-4939-871d-4a54397b376c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_all_probiotic(taxonomy):\n",
    "    \"\"\"Adds probiotic cocktail of AMUC, CBEI, CBUT, BINF, and EHAL to existing taxonomy table at 1% RA each\"\"\"\n",
    "    taxonomy_probiotic_total = pd.DataFrame()\n",
    "    for sample in taxonomy['sample_id'].unique():\n",
    "        taxonomy_reduced = taxonomy[taxonomy['sample_id'] == sample].copy()\n",
    "        taxonomy_reduced['abundance'] = taxonomy_reduced['abundance']*0.95\n",
    "        taxonomy_probiotic = pd.concat([taxonomy_reduced,pd.DataFrame({\n",
    "                  'sample_id':[sample, sample, sample, sample, sample], \n",
    "                  'species': ['Akkermansia muciniphila','Clostridium beijerinckii',\n",
    "                              'Clostridium butyricum','Bifidobacterium longum',\n",
    "                              'Anaerobutyricum hallii'],\n",
    "                  'abundance':[0.01, 0.01, 0.01, 0.01, 0.01], \n",
    "                  'id':['Akkermansia_muciniphila','Clostridium_beijerinckii',\n",
    "                              'Clostridium_butyricum','Bifidobacterium_longum',\n",
    "                              'Anaerobutyricum_hallii']})])\n",
    "        taxonomy_probiotic_total = pd.concat([taxonomy_probiotic_total, taxonomy_probiotic])\n",
    "\n",
    "    return taxonomy_probiotic_total\n",
    "# Add probiotics\n",
    "counts_probiotic = add_all_probiotic(counts_treatment)\n",
    "\n",
    "# Combine duplicate taxa\n",
    "counts_probiotic = counts_probiotic.groupby(['sample_id','id','species']).sum().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef94997c-5d3c-4617-88ab-ec509538f6c0",
   "metadata": {},
   "source": [
    "## Build Models\n",
    "Build MCMMs for samples in the treatment arm and the placebo arm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c820157d-b81e-41f4-8844-0df66bcc19b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build Models\n",
    "agora = '../agora_models/'\n",
    "\n",
    "# Treatment group models\n",
    "micom.workflows.build(counts_probiotic, \n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../WBF011_treated',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n",
    "\n",
    "# Treatment group models w/o probiotics\n",
    "micom.workflows.build(counts_treatment, \n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../WBF011_noProbiotic',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n",
    "\n",
    "# Placebo group models\n",
    "micom.workflows.build(counts_placebo, \n",
    "                                 model_db=agora, \n",
    "                                 out_folder='../Placebo',\n",
    "                                 cutoff=0.001, \n",
    "                                 threads=10)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
